marlowe-playground-client/src/Halogen/ActusBlockly.purs

```diff
@@ -0,0 +1,172 @@
+import Marlowe.ActusBlockly (buildGenerator)
+import Marlowe.Holes (Term(..))
+import Marlowe.Parser as Parser
+import Marlowe.Semantics (Contract(..))
+import Prelude (Unit, bind, const, discard, map, pure, show, unit, zero, ($), (<<<), (<>))
+import Text.Extra as Text
+import Text.Pretty (pretty)
+import Type.Proxy (Proxy(..))
+
+type BlocklyState
+  = { actusBlocklyState :: Maybe BT.BlocklyState
+    , generator :: Maybe Generator
+    , errorMessage :: Maybe String
+    }
+
+_actusBlocklyState :: Lens' BlocklyState (Maybe BT.BlocklyState)
+_actusBlocklyState = prop (SProxy :: SProxy "actusBlocklyState")
+
+_generator :: Lens' BlocklyState (Maybe Generator)
+_generator = prop (SProxy :: SProxy "generator")
+
+_errorMessage :: Lens' BlocklyState (Maybe String)
+_errorMessage = prop (SProxy :: SProxy "errorMessage")
+
+data BlocklyQuery a
+  = Resize a
+  | SetError String a
+
+data ContractFlavour = FS | F
+
+data BlocklyAction
+  = Inject String (Array BlockDefinition)
+  | GetTerms ContractFlavour
+
+data BlocklyMessage
+  = Initialized
+  | CurrentTerms ContractFlavour String
+
+type Slots
+  = ()
+
+type DSL m a
+  = HalogenM BlocklyState BlocklyAction Slots BlocklyMessage m a
+
+blockly :: forall m. MonadEffect m => String -> Array BlockDefinition -> Component HTML BlocklyQuery Unit BlocklyMessage m
+blockly rootBlockName blockDefinitions =
+  mkComponent
+    { initialState: const { actusBlocklyState: Nothing, generator: Nothing, errorMessage: Nothing }
+    , render
+    , eval:
+      H.mkEval
+        { handleQuery
+        , handleAction
+        , initialize: Just $ Inject rootBlockName blockDefinitions
+        , finalize: Nothing
+        , receive: const Nothing
+        }
+    }
+
+handleQuery :: forall m a. MonadEffect m => BlocklyQuery a -> DSL m (Maybe a)
+handleQuery (Resize next) = do
+  mState <- use _actusBlocklyState
+  case mState of
+    Just state ->
+      pure
+        $ ST.run
+            ( do
+                workspaceRef <- STRef.new state.workspace
+                Blockly.resize state.blockly workspaceRef
+            )
+    Nothing -> pure unit
+  pure $ Just next
+
+handleQuery (SetError err next) = do
+  assign _errorMessage $ Just err
+  pure $ Just next
+
+handleAction :: forall m. MonadEffect m => BlocklyAction -> DSL m Unit
+handleAction (Inject rootBlockName blockDefinitions) = do
+  blocklyState <- liftEffect $ Blockly.createBlocklyInstance rootBlockName (ElementId "actusBlocklyWorkspace") (ElementId "actusBlocklyToolbox")
+  let
+    _ =
+      ST.run
+        ( do
+            blocklyRef <- STRef.new blocklyState.blockly
+            workspaceRef <- STRef.new blocklyState.workspace
+            Blockly.addBlockTypes blocklyRef blockDefinitions
+            Blockly.initializeWorkspace blocklyState.blockly workspaceRef
+        )
+
+    generator = buildGenerator blocklyState
+  modify_ _ { actusBlocklyState = Just blocklyState, generator = Just generator }
```

Comment:
what we tend to do with these things is
```haskell
assign _actusBlocklyState (Just blocklyState)
assign _generator (Just generator)
```

Reply:
not that it's better or worse, just keeping consistent

Reply:
Copy paste again :)